packages/server/src/insights/insights.ts

```diff
@@ -0,0 +1,224 @@
+import {
+  QuestionType,
+  Role,
+  PossibleOutputTypes,
+  SimpleDisplayOutputType,
+  SimpleChartOutputType,
+} from '@koh/common';
+import { UserCourseModel } from 'profile/user-course.entity';
+import { SelectQueryBuilder } from 'typeorm';
+import { QuestionModel } from 'question/question.entity';
+
+export interface InsightInterface<Model> {
+  name: string;
+  displayName: string;
+  description: string;
+  roles: Role[];
+  component: string; // In the future we can make this an enum
+  model: new () => Model; // One of the modals have
+  possibleFilters: string[];
+  addFilters: (
+    queryBuilder: SelectQueryBuilder<Model>,
+    filters: any,
+  ) => SelectQueryBuilder<Model>;
+  compute: (
+    queryBuilder: SelectQueryBuilder<Model>,
+    insightFilters: any,
+  ) => Promise<PossibleOutputTypes>;
+  output?: Promise<PossibleOutputTypes>;
+}
+
+class TotalStudents implements InsightInterface<UserCourseModel> {
+  name = 'totalStudents';
+  displayName = 'Total Students';
+  description = 'Gets the total number of students';
+  roles = [Role.PROFESSOR];
+  component = 'SimpleDisplay';
+  model = UserCourseModel;
+  possibleFilters = ['courseId', 'role'];
+
+  async compute(
+    queryBuilder: SelectQueryBuilder<UserCourseModel>,
+    filters,
+  ): Promise<SimpleDisplayOutputType> {
+    return await this.addFilters(
+      queryBuilder.where("role = 'student'"),
+      filters,
+    ).getCount();
+  }
+
+  addFilters(queryBuilder, filters): SelectQueryBuilder<UserCourseModel> {
+    filters.forEach((filter) => {
+      if (!this.possibleFilters.includes(filter.type)) {
+        throw new Error(
+          `${filter} is not a possbile filter for "${this.name}"`,
```

Comment:
why do we have to do this on every insight?

Reply:
Good catch, I'll abstract that.

Reply:
Filters abstracted.
https://github.com/sandboxnu/office-hours/pull/481/commits/360a511d6a6aefcf6d9ef365f84783200d5b6fce

Reply:
the display type is officially now an Enum!!